### PR TITLE
z80: add z84c015 with System Control Registers

### PR DESCRIPTION
**Pull request by holub**
System Control Registers (#ee/ef) for z84c015

### CODE REVIEW

File: src/devices/cpu/z80/tmpz84c015.h [hunk 1]
```diff
@@ -26,8 +26,11 @@ class tmpz84c015_device : public z80_device
 {
 public:
 	tmpz84c015_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t);
+	tmpz84c015_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t, address_map_constructor io_map);
```
Comment (cuavas) [inline src/devices/cpu/z80/tmpz84c015.h:30]: Please make the second constructor here `protected` – it should only be used for derived classes.

File: src/devices/cpu/z80/tmpz84c015.h [hunk 2]
```diff
@@ -26,8 +26,11 @@ class tmpz84c015_device : public z80_device
...
 
 	// configuration helpers
+	template <int Channel> void set_clk_trg(u32 clock) { if (m_ctc) m_ctc->set_clk<Channel>(clock); else subdevice<z80ctc_device>(m_ctc.finder_tag())->set_clk<Channel>(clock); }
+	template <int Channel> void set_clk_trg(const XTAL &xtal) { if (m_ctc) m_ctc->set_clk<Channel>(xtal); else subdevice<z80ctc_device>(m_ctc.finder_tag())->set_clk<Channel>(xtal); }
```
Comment (cuavas) [inline src/devices/cpu/z80/tmpz84c015.h:33]: Is there really any point having the `if` statement here?  The primary case for calling this is during configuration of the parent device where `m_ctc` will be unresolved.  You basically need to call `subdevice`.

File: src/devices/cpu/z80/z84c015.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+//  device_reset - device-specific reset
+//-------------------------------------------------
+void z84c015_device::device_reset()
+{
+	scrp_w(0);
+
+	m_wcr = 0x00; // 0xff, then 0x00 on 16th M1
+    m_mwbr = 0xf0;
+	m_csbr |= 0x0f;
+	m_mcr = 0x01;
```
Comment (cuavas) [inline src/devices/cpu/z80/z84c015.cpp:52]: The high bits of `m_csbr` aren’t initialised here – you should probably initialise them somewhere (probably in `device_start`) for consistent behaviour across runs.

File: src/devices/cpu/z80/tmpz84c015.h (hunk 2 above)
Comment (holub) [reply src/devices/cpu/z80/tmpz84c015.h:33]: I don't have strong opinion here. That was @smf- suggestion.

Comment (cuavas) [reply src/devices/cpu/z80/tmpz84c015.h:33]: Just use the `subdevice` calls – it’ll make the line shorter and easier to read.

File: src/devices/cpu/z80/z84c015.cpp (hunk 3 above)
Comment (holub) [reply src/devices/cpu/z80/z84c015.cpp]: Actually all these PoR, moved to device_start()

**Comment by cuavas** [comment]
Is it intentional that this no longer resets the registers on device reset?  I wasn’t complaining about that, I was just requesting that the other four bits of `m_csbr` that aren't set on device reset should be initialised at some point before that.

**Comment by holub** [comment]
I had rechecked the spec and these are PoR updates. As far as tested reset doesn't break anything for me prefer to keep it this way.

**Comment by cuavas** [comment]
I finally found the documentation for this.  Confusingly, the Toshiba TMPZ84C015 doesn’t have these registers, while that Zilog Z84C015 and Z84C013 do.

If I understand it correctly, the chip is configured to assert CS0 and/or CS1 for specific memory access ranges.  The interface, where the device can be queried for the CS0 or CS1 value for a particular address, doesn’t really reflect how the chip works.

Wouldn’t the right way to do this be to extend the address width for this chip to 18 bits, and expose CS0 and CS1 as high address bits?

@galibert do you have an opinion?

**Comment by holub** [comment]
I made a comment regarding cs implementation in the code `As z80 doesn't implements A lines, addr must be supplied`
Maybe it worth to rename methods to something like `xxx_cs0_r(u16 addr)`?

**Comment by galibert** [comment]
Honestly I'm not sure how these cs* lines should be handled.  I have a similar issues with the tmp68301 where some ranges trigger cs lines (and also wait states).  I have a feeling, since it's ranges, the simplest way is just to log the setup of the cs registers and otherwise handle things as usual through the address map.  I'm not sure extending the address would be useful given it's exclusive anyway.

**Comment by holub** [comment]
anything I can do?

**Comment by holub** [comment]
I merged this to #10754 and moved controversial csN_r(addr) to the driver

**Comment by cuavas** [comment]
How do systems typically use the chip select signals on the Z84C013/Z84C015?  Do they just set it up on initialisation, or do they change it on-the-fly?  If they change it on-the-fly, it effectively *is* used as an extension of the address bus, and could probably be handled by overriding `memory_translate` from `device_memory_interface` to

**Comment by holub** [comment]
Thank you for reviewing!
Systems are changing them on the fly. I will try to reimplement it later based on the suggestion.

**Comment by holub** [comment]
> How do systems typically use the chip select signals on the Z84C013/Z84C015? Do they just set it up on initialisation, or do they change it on-the-fly? If they change it on-the-fly, it effectively _is_ used as an extension of the address bus, and could probably be handled by overriding `memory_translate` from `device_memory_interface` to convert a 16-bit address to an 18-bit address including the chip selects.
> 
> The power-up state enables /CS0 across the entire address range, which would require mapping peripherals into the 0x1xxxx (possibly using mirroring).

I made some progress but unfortunately can't figure out how do mirroring for views. E.g. I need something like:
```
  map(0x0000, 0x3fff).mirror.(0x30000).view(m_bank_view0);
  m_bank_view0[0](0x0000, 0x3fff).mirror.(0x30000).nopw();
  ...
```

**Comment by cuavas** [comment]
I don’t think you can apply `mirror` to a view, only to stuff inside the view.

**Comment by holub** [comment]
> I don’t think you can apply `mirror` to a view, only to stuff inside the view.

I have everything wotking from my end in #11018
Can you take a look at purity of implementation there?

**Pull request closed without merging**